sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/spanner/SpannerIO.java

```diff
@@ -0,0 +1,303 @@
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.List;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ParDo;
+import org.apache.beam.sdk.transforms.display.DisplayData;
+import org.apache.beam.sdk.transforms.display.DisplayData.Builder;
+import org.apache.beam.sdk.util.FluentBackoff;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.PDone;
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+/**
+ * {@link SpannerIO} provides an API for reading from and writing to
+ * <a href="https://cloud.google.com/spanner/">Google Cloud Spanner</a>.
+ *
+ */
+@Experimental(Experimental.Kind.SOURCE_SINK)
+public class SpannerIO {
+
+  private SpannerIO() { }
+
+  @VisibleForTesting
+  public static final int SPANNER_MUTATIONS_PER_COMMIT_LIMIT = 20000;
+
+/*
+ * <p>To write a {@link PCollection} to Spanner, use {@link SpannerIO#writeTo} to
+ * specify the database location for output.
+ *
+ * <p>For example:
+ *
+ * <pre> {@code
+ * // Write data to a table.
+ * PipelineOptions options = PipelineOptionsFactory.fromArgs(args).create();
+ * options.getInput = <CSV text file location>
+ *
+ * Pipeline p = Pipeline.create(options);
+ *   p.apply(TextIO.Read.from(options.getInput()))
+ *       .apply(ParDo.of(new ParseLineFn(tableInfo)))
+ *       .apply(SpannerIO.writeTo(options.getProjectId(),
+ *           options.getInstanceId(), options.getDatabaseId()));
+ *
+ * } </pre>
+ */
+  public static Writer writeTo(String projectId, String instanceId, String databaseId) {
+    return new Writer(projectId, instanceId, databaseId);
+  }
+
+  /**
+   * A {@link PTransform} that writes {@link Mutation} objects to Cloud Spanner.
+   *
+   * @see SpannerIO
+   */
+  public static class Writer extends MutationTransform<Mutation> {
+    /**
+     * Note that {@code projectId} is only {@code @Nullable} as a matter of build order, but if
+     * it is {@code null} at instantiation time, an error will be thrown.
+     */
+    Writer(String projectId, String instanceId, String databaseId) {
+      super(projectId, instanceId, databaseId);
+    }
+
+    /**
+     * Returns a new {@link Write} that writes to the Cloud Spanner for the specified location.
+     */
+    public Writer withLocation(String projectId, String instanceId, String databaseId) {
+      checkNotNull(projectId, "projectId");
+      checkNotNull(instanceId, "instanceId");
+      checkNotNull(databaseId, "databaseId");
+      return new Writer(projectId, instanceId, databaseId);
+    }
+
+  }
+
+
+  /**
+   * A {@link PTransform} that writes mutations to Cloud Spanner
+   *
+   * <b>Note:</b> Only idempotent Cloud Spanner mutation operations (upsert, etc.) should
```

Comment:
Instead of saying (upsert, etc) can you enumerate all the operations that can be used.